test/Microsoft.IdentityModel.Tokens.Saml.Tests/Saml2SecurityTokenHandler.Extensibility.TokenReplay.cs

```diff
@@ -0,0 +1,279 @@
+                    "TokenReplayValidatorCustomExceptionTypeDelegate",
+                    utcNow,
+                    CustomTokenReplayValidationDelegates.TokenReplayValidatorCustomExceptionTypeDelegate,
+                    extraStackFrames: 1)
+                {
+                    // TokenReplayValidationError does not handle the exception type 'CustomSecurityTokenException'
+                    ExpectedException = ExpectedException.SecurityTokenException(
+                        LogHelper.FormatInvariant(
+                            Tokens.LogMessages.IDX10002, // "IDX10002: Unknown exception type returned. Type: '{0}'. Message: '{1}'.";
+                            typeof(CustomSecurityTokenException),
+                            nameof(CustomTokenReplayValidationDelegates.TokenReplayValidatorCustomExceptionTypeDelegate))),
+                    TokenReplayValidationError = new TokenReplayValidationError(
+                        new MessageDetail(
+                            nameof(CustomTokenReplayValidationDelegates.TokenReplayValidatorCustomExceptionTypeDelegate), null),
+                        ValidationFailureType.TokenReplayValidationFailed,
+                        typeof(CustomSecurityTokenException),
+                        new StackFrame("CustomTokenReplayValidationDelegates.cs", 274),
+                        expirationTime)
+                });
+
+                // TokenReplayValidationError : ValidationError, ExceptionType: SecurityTokenReplayDetectedException, inner: CustomSecurityTokenReplayDetectedException
+                theoryData.Add(new TokenReplayExtensibilityTheoryData(
+                    "TokenReplayValidatorThrows",
+                    utcNow,
+                    CustomTokenReplayValidationDelegates.TokenReplayValidatorThrows,
+                    extraStackFrames: 0)
+                {
+                    ExpectedException = new ExpectedException(
+                        typeof(SecurityTokenReplayDetectedException),
+                        string.Format(Tokens.LogMessages.IDX10276),
+                        typeof(CustomSecurityTokenReplayDetectedException)),
+                    TokenReplayValidationError = new TokenReplayValidationError(
+                        new MessageDetail(
+                            string.Format(Tokens.LogMessages.IDX10276), null),
+                        ValidationFailureType.TokenReplayValidatorThrew,
+                        typeof(SecurityTokenReplayDetectedException),
+                        new StackFrame("Saml2SecurityTokenHandler.ValidateToken.Internal.cs", 250),
+                        expirationTime,
+                        new SecurityTokenReplayDetectedException(nameof(CustomTokenReplayValidationDelegates.TokenReplayValidatorThrows))
+                    )
+                });
+                #endregion
+
+                return theoryData;
+            }
+        }
+
+        public class TokenReplayExtensibilityTheoryData : TheoryDataBase
```

Comment:
Do we need to have a specialized class for each token type?
Could Saml2Token and Saml2SecurityTokenHandler be just the base types: SecurityToken, SecurityTokenHandler?